### PR TITLE
Add support for SysListView32 controls with a native UIA implementation

### LINKED ISSUE #15283: dotnet 7, winforms, _get_columnCount is not implemented
### Steps to reproduce:
* create a new winforms project;
* add a listview to a form;
*Add multiple columns to a listview;
* Create a new item and add subItems.

### Actual behavior:

NVDA cannot get information about existing columns

### Expected behavior:

NVDA should be able to fetch information about columns. Narrator and jaws can do it.
### NVDA logs, crash dumps and other attachments:
Traceback (most recent call last):
  File "scriptHandler.pyc", line 295, in executeScript
  File "globalCommands.pyc", line 1415, in script_navigatorObject_firstChild
  File "baseObject.pyc", line 62, in __get__
  File "baseObject.pyc", line 168, in _getPropertyViaCache
  File "NVDAObjects\behaviors.pyc", line 825, in _get_firstChild
  File "NVDAObjects\behaviors.pyc", line 820, in _makeCell
  File "baseObject.pyc", line 62, in __get__
  File "baseObject.pyc", line 168, in _getPropertyViaCache
  File "NVDAObjects\behaviors.pyc", line 791, in _get_childCount
  File "baseObject.pyc", line 62, in __get__
  File "baseObject.pyc", line 168, in _getPropertyViaCache
  File "NVDAObjects\UIA\__init__.pyc", line 1917, in _get_columnCount
NotImplementedError

#### NVDA installed/portable/running from source:
Installed
#### NVDA version:
alpha-28868,a198c9b5 (2023.3.0.28868)
#### Windows version:
Windows 10 22H2 (AMD64) build 19045.3208
#### Name and version of other software in use when reproducing the issue:
An app created using windows forms.
#### Does the issue still occur after restarting your computer?
Yes
#### Have you tried any other versions of NVDA? If so, please report their behaviors.

No
#### If NVDA add-ons are disabled, is your problem still occurring?
Yes
#### Does the issue still occur after you run the COM Registration Fixing Tool in NVDA's tools menu?
Yes
## Additional information
I am attaching a sample project with prebuilt binaries
[WinFormsApp1.zip](https://github.com/nvaccess/nvda/files/12325166/WinFormsApp1.zip)

### PR DESCRIPTION
**Pull request by LeonarddeR**
### Link to issue number:
Fixes #15283 

### Summary of the issue:
.NET 7 Win Forms has a native implementation for SysListView32 controls. This conflicts with us listing SysListView32 as a bad UIA Class name.

### Description of user facing changes
Win forms project in #15283 now works as expected for the most part.

### Description of development approach
Implemented SysListView32 in a new UIA module. Note that UIA seems to expose the following tree by default:
- A list object
- Children for the list items having the ListItem control type, but neither GridItem nor TableItem patterns
- Sub items containing both GridItem and TableItem patterns.

I did the following:

1. Introduced fetching of children text and column headers, mostly based on the Outlook app module
2. Implement position info based on row number of the first child and total row count of parent

### Testing strategy:
1. Tested in the win forms project in #15283 
2. By forcing Good Uia Window, tested in the TweeseCake client as well as in CleanMgr. Note that these still take the IAccessible implementation by default as expected,desired.

### Known issues with pull request:
Row navigation doesn't work in the Win Forms project, but this seems to be an issue in Win Forms not properly implementing SetFocus/InvokePattern.

### Change log entries:
Bug fixes
- Fixed support for SysListView32 controls in Windows Forms applications. (#15283)

### Code Review Checklist:

- [x] Pull Request description:
  - description is up to date
  - change log entries
- [x] Testing:
  - Unit tests
  - System (end to end) tests
  - Manual testing
- [x] API is compatible with existing add-ons.
- [x] Documentation:
  - User Documentation
  - Developer / Technical Documentation
  - Context sensitive help for GUI changes
- [x] UX of all users considered:
  - Speech 
  - Braille
  - Low Vision
  - Different web browsers
  - Localization in other languages / culture than English
- [x] Security precautions taken.

### CODE REVIEW

**Comment by leonardder** [comment]
cc @beqabeqa473

**Comment by leonardder** [comment]
Thanks, I added that now.

**Comment by beqabeqa473** [comment]
Thanks, tested it now.

And the final thing, I don't know if it is a problem, just a bit verbossing, i cannot see that in standard .net framework winforms app, when navigating through list, nvda announces something like
Alice; age 25; Gender Female row 1 1 of 2
Is it normal?
I think it would be better to not announce row information on list items itself, just for columns as this gives difference with IA2 listsItems.

Thanks.

**Comment by leonardder** [comment]
That should also be fixed now. I was mistaken in assuming that the rowNumber property was required for fake table navigation, on list views, but that doesn't seem to be the case.

**Comment by beqabeqa473** [comment]
Thanks. It is now ok for me.

File: source/NVDAObjects/UIA/sysListView32.py [hunk 1]
```diff
@@ -0,0 +1,105 @@
...
+
+from comtypes import COMError
+import config
+from config.configFlags import ReportTableHeaders
+import UIAHandler
+from ..behaviors import RowWithFakeNavigation
+from . import ListItem, UIA
+
+
+def findExtraOverlayClasses(obj, clsList):
```
Comment (seanbudd) [inline source/NVDAObjects/UIA/sysListView32.py:17]: can this get type hints please

File: source/NVDAObjects/UIA/sysListView32.py [hunk 2]
```diff
@@ -0,0 +1,105 @@
...
+		clsList.insert(0, SysListViewList)
+
+
+class SysListViewList(UIA):
+	...
+
+
+class SysListViewItem(RowWithFakeNavigation, ListItem):
+
+	def _get_name(self):
```
Comment (seanbudd) [inline source/NVDAObjects/UIA/sysListView32.py:31]: please add the return type

File: source/NVDAObjects/UIA/sysListView32.py [hunk 3]
```diff
@@ -0,0 +1,105 @@
...
+					columnHeaderTextList.append(columnHeaderItem.currentName)
+			columnHeaderText = " ".join(columnHeaderTextList)
+			if columnHeaderText:
+				text = f"{columnHeaderText} {name}"
+			else:
+				text = name
+			textList.append(text)
+		return "; ".join(textList)
+
+	def _get_indexInParent(self):
```
Comment (seanbudd) [inline source/NVDAObjects/UIA/sysListView32.py:71]: please add the return type

File: source/NVDAObjects/UIA/sysListView32.py [hunk 4]
```diff
@@ -0,0 +1,105 @@
...
+		)
+		val = element.getCachedPropertyValueEx(
+			UIAHandler.UIA.UIA_GridItemRowPropertyId,
+			True
+		)
+		if val == UIAHandler.handler.reservedNotSupportedValue:
+			return super().indexInParent
+		return val
+
+	def _get_positionInfo(self):
```
Comment (seanbudd) [inline source/NVDAObjects/UIA/sysListView32.py:89]: please add the return type

File: source/NVDAObjects/UIA/sysListView32.py [hunk 5]
```diff
@@ -0,0 +1,105 @@
...
+			return super().name
+		textList = []
+		for index in range(cachedChildren.length):
+			e = cachedChildren.getElement(index)
+			name = e.cachedName
+			columnHeaderTextList = []
+			if name and config.conf['documentFormatting']['reportTableHeaders'] in (
+				ReportTableHeaders.ROWS_AND_COLUMNS,
+				ReportTableHeaders.COLUMNS,
+			) and index > 0:
```
Comment (seanbudd) [inline source/NVDAObjects/UIA/sysListView32.py:51]: readibility lint fix

```suggestion
			if (
				name
				and config.conf['documentFormatting']['reportTableHeaders'] in (
					ReportTableHeaders.ROWS_AND_COLUMNS,
					ReportTableHeaders.COLUMNS,
				)
				and index > 0
			):
```

**Review by seanbudd — COMMENTED** [review]
looks good, just some minor changes

File: source/NVDAObjects/UIA/sysListView32.py (hunk 2 above)
Comment (leonardder) [reply source/NVDAObjects/UIA/sysListView32.py:31]: Shouldn't this be inferred from the base?

**Review by seanbudd — APPROVED** [review]
Thanks @leonardder

**Comment by btman16** [comment]
Hi,

This seems to be resulting in extremely sluggish navigation in Task Manager on my laptop.

It now takes over a second to hear the name of the focused item in the list on the Details tab which is an extremely negative hit to productivity, and this is on nVDA 2023.3beta1 where this fix has been implemented. 2023.2 worked fine with the details tab.

This seems to be a pattern with NVDA, whenever something moves away from IAccessible to UIA we then have performance hits.

Thank you very much for your time and I look forward to any input you may have.

Sincerely,

Brandon Tyson

**Comment by leonardder** [comment]
Could you please open a new issue for this? This ensures we have all the details the issue template asks for, such as version of Windows used, etc.

**Pull request merged**